backend/app/analysis/find_vanishing_point.py

```diff
@@ -0,0 +1,227 @@
+
+            # only put line in filter_lines if the angle is NOT within epsilon of
+            # horizontal/vertical
+            epsilon = np.pi / 16
+            if (
+                (point_2_x - point_1_x) != 0
+                and abs(theta - np.pi / 2) > 2 * epsilon
+                and theta > epsilon
+            ):
+                filter_lines.append({
+                    '1_x': int(point_1_x),
+                    '1_y': int(point_1_y),
+                    '2_x': int(point_2_x),
+                    '2_y': int(point_2_y),
+                })
+
+        except ZeroDivisionError:
+            pass
+
+    van_point = find_van_coord_intersections(filter_lines)
+
+    return {
+        'vanishing_point_coord': van_point,
+        'line_coords': filter_lines,
+    }
+
+def auto_canny(image):
+    """
+    Applies the Canny and HoughLinesP functions from OpenCV to the given
+    image and returns the result
+
+    :param image: an image
+    :return: a list containing the binary image from Canny and the set of
+             lines from HoughLinesP
+
+    TODO: Optimize parameters for Canny and HoughLinesP instead of hard-coding them
+    """
+
+    edged = cv2.Canny(image, 150, 255, 3)
+    lines = cv2.HoughLinesP(edged, 1, np.pi / 180, 80, 30, maxLineGap=100)
+
+    # return the edges and lines
+    return [edged, lines]
+
+
+def find_van_coord_intersections(lines):
+    """
+    Find the approximate vanishing point of the image by choosing the most representative point
+    from the largest cluster of intersections between lines in 'lines'
+    :param lines: list of lists containing start and end points of all filtered lines
+    :return: the average coordinate of the largest cluster of intersections
+    """
+    intersections = {}
+    tolerance = 30
+    # print(len(lines))
+    # return None if the image likely has many unnecessary lines (e.g. if there's a tree)
+    if len(lines) > 300 or len(lines) < 2:
+        return None
+
+    # For each pair of lines, finds the intersection and checks to see if it's within the tolerance
+    # from an existing intersection. If so, adds the intersection to the corresponding cluster,
+    # else creates a new cluster.
+    for i in range(len(lines) - 1):
+        for j in range(i + 1, len(lines)):
+            intersection = find_intersection_between_two_lines((lines[i], lines[j]))
+            if intersection is not None:
+                (intX, intY) = intersection
+                found = False
+                for coord in intersections:
+                    distance = ((intX - coord[0]) ** 2 + (intY - coord[1]) ** 2) ** (1 / 2)
+                    if distance < tolerance:
+                        intersections[coord].append(intersection)
+                        found = True
+                if not found:
+                    intersections[intersection] = [intersection]
+    max_frequency = 0
+    van_point_list = []
+    # finds the largest cluster of intersections
```

Comment:
nitpick -- I like to add a new line between chunks of code like this, for ease of reading.